crates/biome_js_analyze/src/lint/nursery/no_secrets.rs

```diff
@@ -155,119 +223,284 @@ struct SensitivePattern {
+
+// Known safe patterns start here
+static BASE64_REGEX: LazyLock<Regex> = LazyLock::new(|| {
+    Regex::new(r"^(?:[A-Za-z0-9+/]{4})*(?:[A-Za-z0-9+/]{2}==|[A-Za-z0-9+/]{3}=)?$").unwrap()
+});
+static URL_REGEX: LazyLock<Regex> =
+    LazyLock::new(|| Regex::new(r"^https?://[a-zA-Z0-9.-]+(/[a-zA-Z0-9./_-]*)?$").unwrap());
+static RELATIVE_PATH_REGEX: LazyLock<Regex> =
+    LazyLock::new(|| Regex::new(r"^(?:\.\./|\./|[a-zA-Z0-9_-]+)/?$").unwrap());
+static UNIX_PATH_REGEX: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"^(/[^/]+)+/?$").unwrap());
+static WINDOWS_PATH_REGEX: LazyLock<Regex> =
+    LazyLock::new(|| Regex::new(r"^[a-zA-Z]:\\(?:[^\\]+\\?)*$").unwrap());
+static EMAIL_REGEX: LazyLock<Regex> =
+    LazyLock::new(|| Regex::new(r"^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$").unwrap());
+static PHONE_REGEX: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"^\+?[1-9]\d{1,14}$").unwrap()); // E.164 format
+
+// Combine all known safe patterns into a single list
+static KNOWN_SAFE_PATTERNS: &[&LazyLock<Regex>] = &[
+    &BASE64_REGEX,
+    &URL_REGEX,
+    &RELATIVE_PATH_REGEX,
+    &UNIX_PATH_REGEX,
+    &WINDOWS_PATH_REGEX,
+    &EMAIL_REGEX,
+    &PHONE_REGEX,
+];
 
-const MIN_PATTERN_LEN: usize = 12;
+fn is_known_safe_pattern(data: &str) -> bool {
+    for pattern in KNOWN_SAFE_PATTERNS {
+        if pattern.is_match(data) {
+            return true;
+        }
+    }
+    false
+}
+
+fn detect_secret(data: &str, entropy_threshold: &u16) -> Option<&'static str> {
+    let tokens: Vec<&str> = split_into_tokens(data);
+
+    for token in tokens {
+        if token.len() >= MIN_PATTERN_LEN {
+            if is_known_safe_pattern(token) {
+                continue;
+            }
 
-fn is_high_entropy(text: &str) -> bool {
-    let entropy = calculate_shannon_entropy(text);
-    entropy > HIGH_ENTROPY_THRESHOLD // TODO: Make this optional, or controllable
+            let entropy =
+                calculate_entropy_with_case_and_classes(token, *entropy_threshold as f64, 15.0);
+            if (entropy as u16) > *entropy_threshold {
+                return Some("Detected high entropy string");
+            }
+        }
+    }
+    None
 }
 
-/// Inspired by https://github.com/nickdeis/eslint-plugin-no-secrets/blob/master/utils.js#L93
-/// Adapted from https://docs.rs/entropy/latest/src/entropy/lib.rs.html#14-33
-/// Calculates Shannon entropy to measure the randomness of data. High entropy values indicate potentially
-/// secret or sensitive information, as such data is typically more random and less predictable than regular text.
-/// Useful for detecting API keys, passwords, and other secrets within code or configuration files.
-fn calculate_shannon_entropy(data: &str) -> f64 {
+/*
+Uses Shannon Entropy as a base algorithm, then adds "boosts" for special patterns/occurrences.
+For example, Continuous mixed cases (lIkE tHiS) are more likely to contribute to a higher score than single cases.
+Symbols also contribute highly to secrets.
```

Comment:
Have you based this on existing works? This could be worth adding references.

Reply:
Not really, I did some prompt engineering (attached in the ChatGPT chat). I'm considering reading some paper(s) on this topic to see if there's a better entropy function out in the wild. Do you have any recommendations for reads?

Reply:
Unfortunately I have no expertise in that domain and I didn't take the time to read the literature. Have you tried the new version of the rule on some code bases?

Reply:
I've added more test cases in valid.js, it was able to clear at least 1-2 more false positives (still leaves some leftover ones). Not a HUGE impact though.

I will see if I can find time to read through the literature, however, I might defer it to one of my company's team members if it takes too long.